manual-provision-payg/testsuite/code_gen.py

```diff
@@ -0,0 +1,320 @@
+        test_code_data["device_id"], json.dumps(test_code_data["test_codes"]), test_code_data["key"]))
+
+    connection_obj.commit()
+    connection_obj.close()
+
+
+def key_gen(key, device_id):
+    "generates 3 test codes to based on installed key. Period currently set to 1h"
+
+    SIGN_WIDTH_BITS = 13
+    COUNTER_WIDTH_BITS = 8
+
+    counter_list = [250, 249, 248]
+    test_code_list = []
+
+    for count in counter_list:
+
+        key_bytes = bytes(key, "utf-8")
+
+        final_code = None
+
+        digest_maker = None
+        digest_maker = hmac.new(key_bytes, None, hashlib.sha1)
+        digest_maker.update(struct.pack("B", period))
+        digest_maker.update(struct.pack("B", count))
+
+        hmac_data = [0]
+        hmac_data = bytearray(digest_maker.digest())
+        # print(base64.b64encode(hmac_data))
+        sign_mask = (1 << SIGN_WIDTH_BITS) - 1
+        sign_result = ((hmac_data[18]) << 8 | (hmac_data[19])) & sign_mask
+
+        code_value = (period << (COUNTER_WIDTH_BITS + SIGN_WIDTH_BITS)
+                      ) | (count << SIGN_WIDTH_BITS) | sign_result
+
+        if len(str(code_value)) < 8:
+            final_code = "0" + str(code_value)
+
+        else:
+            final_code = str(code_value)
+            # print(code_value)
+
+        test_code_list.append(final_code)
+
+    return {"device_id": device_id, "test_codes": test_code_list, "key": key}
+
+
+def write_to_csv(test_code_data):
+    """write test unlock codes to csv file"""
+    fields_exit = False
+
+    try:
+        if 'device_id' in open(cwd+key_code_file_name+".csv").read():
+            fields_exit = True
+        else:
+            fields_exit = False
+
+    except Exception as e:
+        fields_exit = False
+
+    file_wrt = csv.writer(open(cwd+key_code_file_name+".csv", "a"))
+
+    if fields_exit:
+        file_wrt.writerow([test_code_data["device_id"], test_code_data["test_codes"][0],
+                           test_code_data["test_codes"][1], test_code_data["test_codes"][1], test_code_data["key"]])
+    else:
+        file_wrt.writerow(["device_id", "code1", "code2", "code3", "key"])
+        file_wrt.writerow([test_code_data["device_id"], test_code_data["test_codes"][0],
+                           test_code_data["test_codes"][1], test_code_data["test_codes"][2], test_code_data["key"]])
+
+
+def write_to_json(test_code_data):
+    current_paygo_data = []
+
+    try:
+        with open(cwd+key_code_file_name+'.json') as f:
+            current_paygo_data = json.load(f)
+    except:
+        # file doesn't exist
+        pass
+    key_obj = {"device_id": test_code_data["device_id"], "code1": test_code_data["test_codes"][0],
+               "code2": test_code_data["test_codes"][1], "code3": test_code_data["test_codes"][2], "key": test_code_data["key"]}
+    current_paygo_data.append(key_obj)
+
+    with open(cwd+key_code_file_name+'.json', mode='w', encoding='utf-8') as test_codes:
+        json.dump(current_paygo_data, test_codes)
```

Comment:
Why do you need to store this data in JSON as well as CSV?

Reply:
I've found it incredibly useful to have json as a quick machine readable format, especially for upload. Csv's great to view on a spreadsheet to get testcodes. I think we should have both options.